File: flutter/lib/push_handlers.dart
import 'dart:convert';
import 'dart:io';

import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:zixflow/zixflow.dart';

import 'config.dart';

/// Parses the Zixflow `action_buttons` payload (JSON string or list).
List<Map<String, dynamic>> parseActionButtons(dynamic raw) {
  if (raw == null) return [];
  try {
    final decoded = raw is String ? jsonDecode(raw) : raw;
    return List<Map<String, dynamic>>.from(decoded as List);
  } catch (e) {
    debugPrint('[PushHandlers] Error parsing action buttons: $e');
    return [];
  }
}

const String _androidChannelId = 'zixflow_default';
const String _androidChannelName = 'Zixflow Notifications';

/// Background isolate entry point for FCM messages (app backgrounded or
/// terminated). Pure Dart: builds and displays the local notification —
/// including action buttons — directly here, the same way
/// [flutter_local_notifications] is used from a background isolate in any
/// Flutter app. No native Kotlin/Swift code required (mirrors zepto_poc).
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  _logIncomingPush('BACKGROUND/TERMINATED', message);

  // Fresh plugin instance for this isolate — method-channel calls are
  // stateless, so there's no need to share the foreground singleton.
  final plugin = FlutterLocalNotificationsPlugin();
  await plugin.initialize(
    const InitializationSettings(
      android: AndroidInitializationSettings('app_icon'),
    ),
    onDidReceiveNotificationResponse: _notificationTapBackground,
  );
  final androidPlugin = plugin.resolvePlatformSpecificImplementation<
      AndroidFlutterLocalNotificationsPlugin>();
  await androidPlugin?.createNotificationChannel(
    const AndroidNotificationChannel(
      _androidChannelId,
      _androidChannelName,
      description: 'Zixflow push notifications',
      importance: Importance.high,
    ),
  );

  await _showLocalNotification(plugin, message);
}

/// Handles a notification tap (body or action button) delivered to a
/// background isolate, i.e. the app was fully terminated. Re-initializes the
/// Zixflow SDK (required in a fresh isolate) before tracking the tap.
@pragma('vm:entry-point')
Future<void> _notificationTapBackground(NotificationResponse response) async {
  await Zixflow.initialize(
    config: ZixflowConfig(apiKey: AppConfig.zixflowApiKey),
  );
  _handleNotificationResponse(response);
}

/// Firebase Cloud Messaging + flutter_local_notifications with dynamic
/// action buttons — 100% Dart, no native Kotlin/Swift code required.
class PushHandlers {
  PushHandlers._();

  static final FlutterLocalNotificationsPlugin _localNotifications =
      FlutterLocalNotificationsPlugin();

  static String? _fcmToken;

  /// FCM token, exposed for UI display (e.g. copy-to-clipboard).
  static final ValueNotifier<String?> fcmToken = ValueNotifier<String?>(null);

  static const String _iosCategoryId = 'ZX_2BTN';

  /// Call after [Firebase.initializeApp] and [Zixflow.initialize].
  static Future<void> initialize() async {
    FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);

    await FirebaseMessaging.instance
        .setForegroundNotificationPresentationOptions(
      alert: true,
      badge: true,
      sound: true,
    );

    await _requestPermission();
    await _initializeLocalNotifications();
    await _registerToken();
    _setupMessageListeners();
  }

  static Future<void> _requestPermission() async {
    final settings = await FirebaseMessaging.instance.requestPermission(
      alert: true,
      badge: true,
      sound: true,
      provisional: false,
    );
    debugPrint(
      '[PushHandlers] Permission: ${settings.authorizationStatus}',
    );
  }

  static Future<void> _initializeLocalNotifications() async {
    const androidSettings = AndroidInitializationSettings('app_icon');

    final iosSettings = DarwinInitializationSettings(
      requestAlertPermission: false,
      requestBadgePermission: false,
      requestSoundPermission: false,
      notificationCategories: [
        DarwinNotificationCategory(
          _iosCategoryId,
          actions: [
            DarwinNotificationAction.plain('ACTION_0', 'Action 1'),
            DarwinNotificationAction.plain('ACTION_1', 'Action 2'),
          ],
        ),
      ],
    );

    await _localNotifications.initialize(
      InitializationSettings(
        android: androidSettings,
        iOS: iosSettings,
      ),
      onDidReceiveNotificationResponse: _handleNotificationResponse,
      onDidReceiveBackgroundNotificationResponse: _notificationTapBackground,
    );

    final androidPlugin = _localNotifications
        .resolvePlatformSpecificImplementation<
            AndroidFlutterLocalNotificationsPlugin>();
    await androidPlugin?.createNotificationChannel(
      const AndroidNotificationChannel(
        _androidChannelId,
        _androidChannelName,
        description: 'Zixflow push notifications',
        importance: Importance.high,
      ),
    );
  }

  static Future<void> _registerToken() async {
    _fcmToken = await FirebaseMessaging.instance.getToken();
    if (_fcmToken != null) {
      fcmToken.value = _fcmToken;
      Zixflow.instance.registerDeviceToken(deviceToken: _fcmToken!);
      _printToken('FCM token registered', _fcmToken!);
    }

    FirebaseMessaging.instance.onTokenRefresh.listen((newToken) {
      _fcmToken = newToken;
      fcmToken.value = newToken;
      Zixflow.instance.registerDeviceToken(deviceToken: newToken);
      _printToken('FCM token refreshed', newToken);
    });
  }

  static void _printToken(String label, String token) {
    debugPrint('');
    debugPrint('════════════════════════════════════════');
    debugPrint('[PushHandlers] $label:');
    debugPrint(token);
    debugPrint('════════════════════════════════════════');
    debugPrint('');
  }

  static void _setupMessageListeners() {
    // Foreground: show the local notification ourselves (with buttons).
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      _logIncomingPush('FOREGROUND', message);
      _showLocalNotification(_localNotifications, message);
    });

    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      _logIncomingPush('OPENED (tapped from background)', message);
      _trackOpened(message.data);
      _handleDeeplink(message.data['deeplink_url']?.toString());
    });

    FirebaseMessaging.instance.getInitialMessage().then((message) {
      if (message != null) {
        _logIncomingPush('OPENED (launched from terminated)', message);
        _trackOpened(message.data);
        _handleDeeplink(message.data['deeplink_url']?.toString());
      }
    });
  }
}

/// Handles taps on local notifications (body or ACTION_0 / ACTION_1) from the
/// foreground/running-app isolate. Shared by [PushHandlers] and the
/// background entry point.
void _handleNotificationResponse(NotificationResponse response) {
  if (response.payload == null || response.payload!.isEmpty) return;

  Map<String, dynamic> payload;
  try {
    payload = jsonDecode(response.payload!) as Map<String, dynamic>;
  } catch (_) {
    return;
  }

  _trackOpened(payload);

  if (response.actionId != null && response.actionId!.isNotEmpty) {
    _trackActionClick(payload, response.actionId!);
    final buttons = parseActionButtons(payload['action_buttons']);
    final actionIndex =
        int.tryParse(response.actionId!.replaceAll(RegExp(r'\D'), '')) ?? -1;
    final buttonDeeplink = (actionIndex >= 0 && actionIndex < buttons.length)
        ? buttons[actionIndex]['deeplink']?.toString() ?? ''
        : '';
    _handleDeeplink(
      buttonDeeplink.isNotEmpty
          ? buttonDeeplink
          : payload['deeplink_url']?.toString(),
    );
  } else {
    _handleDeeplink(payload['deeplink_url']?.toString());
  }
}

void _trackOpened(Map<String, dynamic> data) {
  final deliveryId = data['Zixflow-Delivery-ID']?.toString() ?? '';
  final deliveryToken = data['Zixflow-Delivery-Token']?.toString() ?? '';

  if (deliveryId.isNotEmpty && deliveryToken.isNotEmpty) {
    Zixflow.instance.trackMetric(
      deliveryID: deliveryId,
      deviceToken: deliveryToken,
      event: MetricEvent.opened,
    );
  }
}

void _trackActionClick(Map<String, dynamic> payload, String actionId) {
  final actionIndex =
      int.tryParse(actionId.replaceAll(RegExp(r'\D'), '')) ?? -1;
  final buttons = parseActionButtons(payload['action_buttons']);
  final actionName = (actionIndex >= 0 && actionIndex < buttons.length)
      ? buttons[actionIndex]['name']?.toString() ?? 'Action ${actionIndex + 1}'
      : 'Action ${actionIndex + 1}';
  final actionDeeplink = (actionIndex >= 0 && actionIndex < buttons.length)
      ? buttons[actionIndex]['deeplink']?.toString() ?? ''
      : '';

  Zixflow.instance.track(
    name: 'Push Notification Action Clicked',
    properties: {
      'Zixflow-Delivery-ID': payload['Zixflow-Delivery-ID'] ?? '',
      'Zixflow-Delivery-Token': payload['Zixflow-Delivery-Token'] ?? '',
      'notification_id': payload['Zixflow-Delivery-ID'] ?? '',
      'title': payload['title'] ?? '',
      'action_id': actionId,
      'action_index': actionIndex,
      'action_name': actionName,
      'action_deeplink': actionDeeplink,
      'source': 'local_notification',
    },
  );
}

/// Downloads [url] to a temp file and returns its local path, or null on
/// failure / when [url] is missing or not http(s).
Future<String?> _downloadToTempFile(String? url, String fileName) async {
  if (url == null || url.isEmpty || !url.startsWith('http')) return null;
  try {
    final response = await http.get(Uri.parse(url));
    if (response.statusCode != 200) return null;
    final dir = await getTemporaryDirectory();
    final file = File('${dir.path}/$fileName');
    await file.writeAsBytes(response.bodyBytes);
    return file.path;
  } catch (e) {
    debugPrint('[PushHandlers] Failed to download $url: $e');
    return null;
  }
}

/// Builds and displays the local notification (with dynamic action buttons,
/// rich media, sound, and badge) for the given [message] using the provided
/// plugin instance. Works from both the main isolate (foreground) and a
/// background isolate.
Future<void> _showLocalNotification(
  FlutterLocalNotificationsPlugin plugin,
  RemoteMessage message,
) async {
  final data = Map<String, dynamic>.from(message.data);
  final title =
      message.notification?.title ?? data['title']?.toString() ?? 'Notification';
  final body = message.notification?.body ?? data['body']?.toString() ?? '';

  // Prefer notification title/body in the payload for action-click tracking.
  data['title'] ??= title;
  data['body'] ??= body;

  final imageUrl = data['image_url']?.toString();
  final largeIconUrl = data['large_icon_url']?.toString();
  final soundName = data['sound']?.toString();
  final badgeNumber = int.tryParse(data['badge']?.toString() ?? '');
  final id = message.hashCode;

  // Download rich media concurrently (Android BigPictureStyle / iOS attachment).
  final imagePath = await _downloadToTempFile(imageUrl, 'push_img_$id.jpg');
  final largeIconPath =
      await _downloadToTempFile(largeIconUrl, 'push_icon_$id.jpg');

  final buttons = parseActionButtons(data['action_buttons']);
  final androidActions = <AndroidNotificationAction>[];
  for (var i = 0; i < buttons.length && i < 2; i++) {
    androidActions.add(
      AndroidNotificationAction(
        'ACTION_$i',
        buttons[i]['name']?.toString() ?? 'Action ${i + 1}',
        showsUserInterface: true,
      ),
    );
  }

  StyleInformation? style;
  if (imagePath != null) {
    style = BigPictureStyleInformation(
      FilePathAndroidBitmap(imagePath),
      largeIcon:
          largeIconPath != null ? FilePathAndroidBitmap(largeIconPath) : null,
      contentTitle: title,
      summaryText: body,
      hideExpandedLargeIcon: false,
    );
  }

  AndroidNotificationSound? androidSound;
  if (soundName != null && soundName != 'default' && soundName != 'none') {
    androidSound = RawResourceAndroidNotificationSound(soundName);
  }

  final androidDetails = AndroidNotificationDetails(
    _androidChannelId,
    _androidChannelName,
    channelDescription: 'Zixflow push notifications',
    importance: Importance.high,
    priority: Priority.high,
    actions: androidActions,
    styleInformation: style,
    largeIcon:
        largeIconPath != null ? FilePathAndroidBitmap(largeIconPath) : null,
    playSound: androidSound != null,
    sound: androidSound,
  );

  final iosAttachments = <DarwinNotificationAttachment>[
    if (imagePath != null) DarwinNotificationAttachment(imagePath),
  ];

  final iosDetails = DarwinNotificationDetails(
    presentAlert: true,
    presentBadge: true,
    presentSound: true,
    categoryIdentifier: 'ZX_2BTN',
    attachments: iosAttachments,
    badgeNumber: badgeNumber,
    sound: (soundName != null && soundName != 'default' && soundName != 'none')
        ? soundName
        : null,
  );

  await plugin.show(
    id,
    title,
    body,
    NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    ),
    payload: jsonEncode(data),
  );
}

/// Opens [deeplink] in the browser / external app (works for both `https://`
/// URLs and custom schemes like `zixflow://`).
void _handleDeeplink(String? deeplink) {
  if (deeplink == null || deeplink.isEmpty) return;
  debugPrint('[PushHandlers] Deeplink: $deeplink');
  final uri = Uri.tryParse(deeplink);
  if (uri == null) return;
  launchUrl(uri, mode: LaunchMode.externalApplication).catchError((e) {
    debugPrint('[PushHandlers] Failed to launch deeplink $deeplink: $e');
    return false;
  });
}

/// Logs the full incoming push payload (all RemoteMessage fields + data map)
/// for any app state — foreground, background, or opened.
void _logIncomingPush(String state, RemoteMessage message) {
  const encoder = JsonEncoder.withIndent('  ');
  final title = message.notification?.title ?? message.data['title'] ?? '(no title)';
  final body = message.notification?.body ?? message.data['body'] ?? '(no body)';

  debugPrint('');
  debugPrint('════════════════════════════════════════');
  debugPrint('🔔 PUSH RECEIVED [$state]');
  debugPrint('   messageId       : ${message.messageId}');
  debugPrint('   messageType     : ${message.messageType}');
  debugPrint('   senderId        : ${message.senderId}');
  debugPrint('   category        : ${message.category}');
  debugPrint('   collapseKey     : ${message.collapseKey}');
  debugPrint('   contentAvailable: ${message.contentAvailable}');
  debugPrint('   sentTime        : ${message.sentTime}');
  debugPrint('   ttl             : ${message.ttl}');
  debugPrint('   from            : ${message.from}');
  debugPrint('   title           : $title');
  debugPrint('   body            : $body');
  if (message.notification?.android != null) {
    final android = message.notification!.android!;
    debugPrint('   android.channelId  : ${android.channelId}');
    debugPrint('   android.imageUrl   : ${android.imageUrl}');
    debugPrint('   android.clickAction: ${android.clickAction}');
  }
  if (message.notification?.apple != null) {
    final apple = message.notification!.apple!;
    debugPrint('   apple.badge  : ${apple.badge}');
    debugPrint('   apple.sound  : ${apple.sound?.name}');
    debugPrint('   apple.imageUrl: ${apple.imageUrl}');
  }
  debugPrint('   data (full payload):');
  try {
    encoder.convert(message.data).split('\n').forEach((line) {
      debugPrint('     $line');
    });
  } catch (_) {
    debugPrint('     ${message.data}');
  }
  debugPrint('════════════════════════════════════════');
  debugPrint('');
}

File: react-native/src/pushHandlers.ts
import { Linking, Platform } from 'react-native';
import messaging, {
  FirebaseMessagingTypes,
} from '@react-native-firebase/messaging';
import notifee, {
  AndroidImportance,
  AndroidStyle,
  EventType,
  type Notification,
} from '@notifee/react-native';
import { MetricEvent, Zixflow } from 'zixflow-reactnative';

import { parseActionButtons } from './pushActions';
import { navigate, resolveInAppRoute } from './navigation';

/**
 * Pure JavaScript/TypeScript push notification handling for Android — no
 * custom native Kotlin code required (mirrors the Flutter SDK example, which
 * uses `firebase_messaging` + `flutter_local_notifications` entirely in Dart).
 *
 * `zixflow-reactnative`'s JS API does not emit incoming push messages to JS
 * (only `onMessageReceived` to *report* an already-known message for
 * tracking), so `@react-native-firebase/messaging` is used here as the FCM
 * listener, and `@notifee/react-native` displays the notification — including
 * dynamic action buttons parsed from the `action_buttons` payload field.
 *
 * The Zixflow SDK's own native FCM service still runs in parallel (Android
 * allows multiple manifest-declared FCM receivers) and continues to handle
 * device-token registration/refresh and default delivery tracking; this
 * module owns notification *display* and action-button tracking.
 */

const ANDROID_CHANNEL_ID = 'zixflow_default';
const ANDROID_CHANNEL_NAME = 'Zixflow Notifications';

let cachedToken: string | undefined;

type NotificationDataValue = string | number | object;
type NotificationData = Record<string, NotificationDataValue>;

function toLogString(value: unknown): string {
  if (typeof value === 'string') return value;
  if (value == null) return '';
  try {
    return JSON.stringify(value);
  } catch {
    return '[unserializable]';
  }
}

function sanitizeNotificationData(
  data: Record<string, unknown>,
): NotificationData {
  const entries = Object.entries(data).filter(([, value]) => value !== undefined);
  return Object.fromEntries(entries) as NotificationData;
}

function getStringField(
  data: Record<string, unknown>,
  key: string,
): string | undefined {
  const value = data[key];
  return typeof value === 'string' ? value : undefined;
}

/** Parses the Zixflow `action_buttons` JSON field into notifee actions. */
function buildNotifeeActions(data: Record<string, unknown>) {
  const buttons = parseActionButtons(getStringField(data, 'action_buttons'));
  return buttons.slice(0, 2).map((button, index) => ({
    title: button.name || `Action ${index + 1}`,
    pressAction: { id: `action_${index}` },
  }));
}

/** Logs the full incoming push payload (all RemoteMessage fields + data map). */
function logIncomingPush(
  state: string,
  message: FirebaseMessagingTypes.RemoteMessage,
) {
  const titleRaw =
    message.notification?.title ?? message.data?.title ?? '(no title)';
  const bodyRaw =
    message.notification?.body ?? message.data?.body ?? '(no body)';

  console.log('');
  console.log('════════════════════════════════════════');
  console.log(`🔔 PUSH RECEIVED [${state}]`);
  console.log(`   messageId    : ${message.messageId}`);
  console.log(`   sentTime     : ${message.sentTime}`);
  console.log(`   ttl          : ${message.ttl}`);
  console.log(`   from         : ${message.from}`);
  console.log(`   collapseKey  : ${message.collapseKey}`);
  console.log(`   title        : ${toLogString(titleRaw)}`);
  console.log(`   body         : ${toLogString(bodyRaw)}`);
  console.log('   data (full payload):');
  console.log(JSON.stringify(message.data ?? {}, null, 2));
  console.log('════════════════════════════════════════');
  console.log('');
}

/** Returns true if `value` looks like a usable http(s) image URL for notifee. */
function isValidImageUrl(value?: string): value is string {
  return Boolean(value) && /^https?:\/\//i.test(value as string);
}

/** Displays a local notification (with dynamic action buttons) via notifee. */
async function showNotification(
  message: FirebaseMessagingTypes.RemoteMessage,
) {
  const data: NotificationData = sanitizeNotificationData(
    (message.data ?? {}) as Record<string, unknown>,
  );
  const titleFromData = getStringField(data, 'title');
  const bodyFromData = getStringField(data, 'body');
  const soundFromData = getStringField(data, 'sound');
  const badgeFromData = getStringField(data, 'badge');
  const stickyFromData = getStringField(data, 'sticky');
  const largeIconUrl = getStringField(data, 'large_icon_url');
  const imageUrl = getStringField(data, 'image_url');
  const title =
    message.notification?.title ?? titleFromData ?? 'Notification';
  const body = message.notification?.body ?? bodyFromData ?? '';

  const soundName =
    soundFromData && soundFromData !== 'default' && soundFromData !== 'none'
      ? soundFromData
      : undefined;
  const badgeCount =
    badgeFromData != null ? Number.parseInt(badgeFromData, 10) : undefined;
  // "sticky": true means the notification survives swipe-dismiss and "Clear all", but
  // STILL gets removed when tapped or when an action button is pressed — sticky only
  // blocks passive dismissal, not active interaction. `autoCancel` is therefore always
  // true below; `ongoing` is what's actually driven by `sticky`. Android only — no iOS
  // equivalent.
  const sticky = stickyFromData === 'true';

  if (badgeCount != null && !Number.isNaN(badgeCount)) {
    notifee.setBadgeCount(badgeCount).catch(() => {});
  }

  const hasBadge = badgeCount != null && !Number.isNaN(badgeCount);

  // notifee's validators check for key *presence* (`hasOwnProperty`), not just
  // truthiness — so `sound: undefined` / `badgeCount: undefined` / `largeIcon:
  // undefined` still fail validation ("must be a string/number value if
  // specified") because the key exists on the object. Conditionally spread
  // each optional field so the key is omitted entirely when there's no value.
  await notifee.displayNotification({
    title,
    body,
    data,
    android: {
      channelId: ANDROID_CHANNEL_ID,
      importance: AndroidImportance.HIGH,
      pressAction: { id: 'default' },
      actions: buildNotifeeActions(data),
      ...(isValidImageUrl(largeIconUrl)
        ? { largeIcon: largeIconUrl }
        : {}),
      ...(isValidImageUrl(imageUrl)
        ? { style: { type: AndroidStyle.BIGPICTURE, picture: imageUrl } }
        : {}),
      ...(soundName ? { sound: soundName } : {}),
      autoCancel: true,
      ongoing: sticky,
    },
    ios: {
      categoryId: 'ZX_2BTN',
      ...(soundName ? { sound: soundName } : {}),
      ...(hasBadge ? { badgeCount } : {}),
    },
  });
}


/** Tracks the "opened" metric for a delivered push using its delivery IDs. */
function trackOpened(data: Record<string, unknown>) {
  const deliveryId = getStringField(data, 'Zixflow-Delivery-ID') ?? '';
  const deliveryToken =
    getStringField(data, 'Zixflow-Delivery-Token') ?? cachedToken ?? '';

  if (deliveryId && deliveryToken) {
    Zixflow.trackMetric({
      deliveryID: deliveryId,
      deviceToken: deliveryToken,
      event: MetricEvent.Opened,
    }).catch(() => {});
  }
}

/** Tracks the "delivered" metric as soon as the data payload arrives — call this
 * from both the foreground `onMessage` listener and the background handler,
 * before displaying the local notification. */
function trackDelivered(data: Record<string, unknown>) {
  const deliveryId = getStringField(data, 'Zixflow-Delivery-ID') ?? '';
  const deliveryToken = getStringField(data, 'Zixflow-Delivery-Token') ?? cachedToken ?? '';

  if (deliveryId && deliveryToken) {
    Zixflow.trackMetric({
      deliveryID: deliveryId,
      deviceToken: deliveryToken,
      event: MetricEvent.Delivered,
    }).catch(() => {});
  }
}

/** Tracks "Push Notification Action Clicked" and opens the button's deeplink. */
function trackActionClick(
  data: Record<string, unknown>,
  actionId: string,
) {
  const actionIndex = Number.parseInt(actionId.replace(/\D/g, ''), 10) || 0;
  const buttons = parseActionButtons(getStringField(data, 'action_buttons'));
  const button = buttons[actionIndex];
  const actionName = button?.name || `Action ${actionIndex + 1}`;
  const actionDeeplink = button?.deeplink ?? '';

  Zixflow.track('Push Notification Action Clicked', {
    'Zixflow-Delivery-ID': getStringField(data, 'Zixflow-Delivery-ID') ?? '',
    'Zixflow-Delivery-Token': getStringField(data, 'Zixflow-Delivery-Token') ?? '',
    notification_id: getStringField(data, 'Zixflow-Delivery-ID') ?? '',
    title: getStringField(data, 'title') ?? '',
    action_id: actionId,
    action_index: actionIndex,
    action_name: actionName,
    action_deeplink: actionDeeplink,
    source: 'local_notification',
  }).catch(() => {});

  handleDeeplink(actionDeeplink || getStringField(data, 'deeplink_url'));
}

function handleDeeplink(deeplink?: string) {
  if (!deeplink) return;
  console.log(`[PushHandlers] Deeplink: ${deeplink}`);

  const screen = resolveInAppRoute(deeplink);
  if (screen) {
    navigate(screen);
    return;
  }

  Linking.openURL(deeplink).catch(() => {});
}

/** Handles a notifee event (foreground or background) for a tap or action press. */
function handleNotifeeEvent(type: EventType, detail: { notification?: Notification; pressAction?: { id: string } }) {
  const data = (detail.notification?.data ?? {}) as Record<string, unknown>;

  if (type === EventType.PRESS) {
    trackOpened(data);
    handleDeeplink(data.deeplink_url);
  } else if (type === EventType.ACTION_PRESS && detail.pressAction) {
    trackOpened(data);
    trackActionClick(data, detail.pressAction.id);
    // Action buttons on Android are routed through notifee's background handler, so
    // `autoCancel` does NOT dismiss the notification for action presses (only for the
    // default body-tap `pressAction`) — it must be cancelled explicitly. This applies
    // regardless of `sticky`: pressing a button always removes the notification; only
    // swipe/"Clear all" is blocked when the push was marked sticky (via `ongoing`).
    if (detail.notification?.id) {
      notifee.cancelNotification(detail.notification.id).catch(() => {});
    }
  }
}

export const PushHandlers = {
  /** FCM token, updated on registration/refresh — read by the UI for display. */
  fcmToken: undefined as string | undefined,

  /** Call after `Zixflow.initialize()`. Sets up permissions, token, and listeners. */
  async initialize(): Promise<void> {
    await notifee.requestPermission();
    if (Platform.OS === 'android') {
      await notifee.createChannel({
        id: ANDROID_CHANNEL_ID,
        name: ANDROID_CHANNEL_NAME,
        importance: AndroidImportance.HIGH,
      });
    }

    if (Platform.OS === 'ios') {
      // iOS requires explicit APNs registration + permission via
      // `registerDeviceForRemoteMessages()` before `getToken()` will resolve.
      if (!messaging().isDeviceRegisteredForRemoteMessages) {
        await messaging().registerDeviceForRemoteMessages();
      }
    }

    await registerToken();
    messaging().onTokenRefresh(async (newToken) => {
      cachedToken = newToken;
      PushHandlers.fcmToken = newToken;
      await Zixflow.registerDeviceToken(newToken);
      console.log(`[PushHandlers] FCM token refreshed: ${newToken}`);
    });

    // Foreground: FCM message received while app is open.
    messaging().onMessage(async (message) => {
      logIncomingPush('FOREGROUND', message);
      trackDelivered((message.data ?? {}) as Record<string, unknown>);
      await showNotification(message);
    });

    // Notification tap that brought the app from background to foreground.
    messaging().onNotificationOpenedApp((message) => {
      logIncomingPush('OPENED (tapped from background)', message);
      trackOpened((message.data ?? {}) as Record<string, unknown>);
      handleDeeplink(message.data?.deeplink_url as string | undefined);
    });

    // App launched by tapping a notification while fully terminated.
    const initialMessage = await messaging().getInitialNotification();
    if (initialMessage) {
      logIncomingPush('OPENED (launched from terminated)', initialMessage);
      trackOpened((initialMessage.data ?? {}) as Record<string, unknown>);
      handleDeeplink(initialMessage.data?.deeplink_url as string | undefined);
    }

    // Foreground taps/action presses on notifee-displayed notifications.
    notifee.onForegroundEvent(({ type, detail }) => {
      handleNotifeeEvent(type, detail);
    });
  },
};

async function registerToken() {
  try {
    const token = await messaging().getToken();
    cachedToken = token;
    PushHandlers.fcmToken = token;
    await Zixflow.registerDeviceToken(token);
    console.log('');
    console.log('════════════════════════════════════════');
    console.log('[PushHandlers] FCM token registered:');
    console.log(token);
    console.log('════════════════════════════════════════');
    console.log('');
  } catch (error) {
    console.log(`[PushHandlers] Failed to register token: ${error}`);
  }
}

/**
 * Background FCM message handler — must be registered at the top level of
 * `index.js` (before `AppRegistry.registerComponent`), same pattern as
 * `FirebaseMessaging.onBackgroundMessage` in Flutter. Displays the
 * notification (with action buttons) while the app is backgrounded/killed.
 */
export async function firebaseBackgroundMessageHandler(
  message: FirebaseMessagingTypes.RemoteMessage,
): Promise<void> {
  logIncomingPush('BACKGROUND/TERMINATED', message);
  trackDelivered((message.data ?? {}) as Record<string, unknown>);
  await showNotification(message);
}

/**
 * Background notifee event handler (action button / notification tap while
 * app is backgrounded or terminated) — must be registered at the top level of
 * `index.js`.
 */
export async function notifeeBackgroundEventHandler({
  type,
  detail,
}: {
  type: EventType;
  detail: { notification?: Notification; pressAction?: { id: string } };
}): Promise<void> {
  handleNotifeeEvent(type, detail);
}
